cmd/dockerd/config.go

```diff
@@ -28,6 +28,7 @@ func installCommonConfigFlags(conf *config.Config, flags *pflag.FlagSet) {
 	flags.StringVar(&conf.ExecRoot, "exec-root", conf.ExecRoot, "Root directory for execution state files")
 	flags.StringVar(&conf.ContainerdAddr, "containerd", "", "containerd grpc address")
 	flags.BoolVar(&conf.CriContainerd, "cri-containerd", false, "start containerd with cri")
+	flags.Var(opts.NewNamedSetOpts("feature", conf.Features), "feature", "Enable feature in the daemon")
```

Comment:
Oh (haven't  checked yet); also wondering now what we do with other options; if we accept a comma-separated list, or only multiple flags (i.e. `--something foo=bar,bar=baz` versus `--something foo=bar --something bar=baz`).

I _THINK_ we may be using the latter, and because of that are using _singular_ for the flag-names (because each flag takes a single value). For example the `--storage-opt` flag only takes a single option, but the equivalent in the config-file is `storage-opts` (plural) as it accepts multiple.

We may need to look at consistency there.

Reply:
That's already how this is implemented, right?  Adding a new `--feature` flag which, when specified multiple times, adds entries to the (existing) `features` map in the configuration structure.

Reply:
Ah, you're right, I think I mis-interpreted the type itself to accepting multiple values.

But this one should use plural (`features`); the name here is used to correlate flags with the config (`daemon.json`) to detect conflicts.

Currently, this PR doesn't detect those:

```bash
echo '{"features": {"containerd-snapshotter": true}}' > /etc/docker/daemon.json

hack/make.sh binary && make install
dockerd --feature containerd-snapshotter=true --validate
configuration OK
```

Changing this to plural;

```patch
diff --git a/cmd/dockerd/config.go b/cmd/dockerd/config.go
index 2ea669f0b1..bb6e44ac6e 100644
--- a/cmd/dockerd/config.go
+++ b/cmd/dockerd/config.go
@@ -28,7 +28,7 @@ func installCommonConfigFlags(conf *config.Config, flags *pflag.FlagSet) {
 	flags.StringVar(&conf.ExecRoot, "exec-root", conf.ExecRoot, "Root directory for execution state files")
 	flags.StringVar(&conf.ContainerdAddr, "containerd", "", "containerd grpc address")
 	flags.BoolVar(&conf.CriContainerd, "cri-containerd", false, "start containerd with cri")
-	flags.Var(opts.NewNamedSetOpts("feature", conf.Features), "feature", "Enable feature in the daemon")
+	flags.Var(opts.NewNamedSetOpts("features", conf.Features), "feature", "Enable feature in the daemon")

 	flags.Var(opts.NewNamedMapMapOpts("default-network-opts", conf.DefaultNetworkOpts, nil), "default-network-opt", "Default network options")
 	flags.IntVar(&conf.MTU, "mtu", conf.MTU, `Set the MTU for the default "bridge" network`)
```

Then it correctly handles the conflict;

```bash
hack/make.sh binary && make install
dockerd --feature containerd-snapshotter=true --validate
unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives are specified both as a flag and in the configuration file: features: (from flag: map[containerd-snapshotter:true], from file: map[containerd-snapshotter:true])
```